internal/webhook/v1alpha1/firewallrule_webhook.go

```diff
@@ -0,0 +1,133 @@
+func (v *FirewallRuleCustomValidator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	firewallrule, ok := obj.(*v1alpha1.FirewallRule)
+	if !ok {
+		return nil, fmt.Errorf("expected a FirewallRule object but got %T", obj)
+	}
+	firewallrulelog.Info("Validation for FirewallRule upon creation", "name", firewallrule.GetName())
+
+	return v.validate(ctx, firewallrule)
+}
+
+// ValidateUpdate implements webhook.CustomValidator so a webhook will be registered for the type FirewallRule.
+func (v *FirewallRuleCustomValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	firewallrule, ok := newObj.(*v1alpha1.FirewallRule)
+	if !ok {
+		return nil, fmt.Errorf("expected a FirewallRule object for the newObj but got %T", newObj)
+	}
+	firewallrulelog.Info("Validation for FirewallRule upon update", "name", firewallrule.GetName())
+
+	return v.validate(ctx, firewallrule)
+}
+
+// ValidateDelete implements webhook.CustomValidator so a webhook will be registered for the type FirewallRule.
+func (v *FirewallRuleCustomValidator) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	firewallrule, ok := obj.(*v1alpha1.FirewallRule)
+	if !ok {
+		return nil, fmt.Errorf("expected a FirewallRule object but got %T", obj)
+	}
+	firewallrulelog.Info("Validation for FirewallRule upon deletion", "name", firewallrule.GetName())
+
+	// TODO(user): fill in your validation logic upon object deletion.
+
+	return nil, nil
+}
+
+// validate implements webhook.CustomValidator so a webhook will be registered for the type FirewallRule.
+func (v *FirewallRuleCustomValidator) validate(ctx context.Context, firewallrule *v1alpha1.FirewallRule) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+
+	path := field.NewPath("spec")
+	nodename := ptr.Deref(firewallrule.Spec.NodeName, "")
+	if nodename != "" {
+		var firewallrules v1alpha1.FirewallRuleList
+
+		if err := v.List(ctx, &firewallrules, client.MatchingFields{firewallRuleNodeNameKey: nodename}); err != nil {
+			firewallrulelog.Error(err, "Unable to list FirewallRules")
+			return nil, err
+		}
+
+		firewallrules.Items = slices.DeleteFunc(firewallrules.Items, func(fr v1alpha1.FirewallRule) bool {
+			return fr.Name == firewallrule.Name
+		})
+
+		frSpec := provider.EncodeFirewallRuleSpec(firewallrule)
+		frSpecs := provider.EncodeFirewallRuleSpecs(firewallrules.Items)
+
+		if frSpec.Direction == provider.DirectionIngress &&
+			provider.ContainsPermission(provider.GetIngressIPPermissions(frSpecs), frSpec.IPPermission) {
+			allErrs = append(allErrs, field.Invalid(path, firewallrule.Spec, "ingress rule already exists."))
+		} else if frSpec.Direction == provider.DirectionEgress &&
+			provider.ContainsPermission(provider.GetEgressIPPermissions(frSpecs), frSpec.IPPermission) {
+			allErrs = append(allErrs, field.Invalid(path, firewallrule.Spec, "egress rule already exists."))
+		}
```

Comment:
Maybe this type of error would be better ?
```suggestion
		if frSpec.Direction == provider.DirectionIngress &&
			provider.ContainsPermission(provider.GetIngressIPPermissions(frSpecs), frSpec.IPPermission) {
			allErrs = append(allErrs, field.Duplicate(path, firewallrule.Spec))
		} else if frSpec.Direction == provider.DirectionEgress &&
			provider.ContainsPermission(provider.GetEgressIPPermissions(frSpecs), frSpec.IPPermission) {
			allErrs = append(allErrs, field.Duplicate(path, firewallrule.Spec))
		}
```